thrift/src/main/java/com/linecorp/armeria/server/thrift/THttpHeaderInjectingService.java

```diff
@@ -0,0 +1,146 @@
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.SimpleDecoratingService;
+
+/**
+ * A {@link SimpleDecoratingService} that injects {@link HttpHeaders} when using {@link THttpService}.
+ * A Thrift call replies with an {@link RpcResponse} which is wrapped by an {@link HttpResponse}.
+ * The {@link HttpHeaders} returned from {@link #httpHeaders(ServiceRequestContext, RpcRequest, RpcResponse)}
+ * will be added to that {@link HttpResponse}.
+ *
+ * <p>{@link THttpService} should be decorated after this class to work properly. For example:
+ *
+ * <pre>{@code
+ * final ServerBuilder sb = ...
+ * sb.service("/", ThriftCallService.of(thriftHandler)
+ *                                  .decorate(...)
+ *                                   // A subclass of THttpHeaderInjectingService
+ *                                  .decorate(MyInjectingService::new)
+ *                                  .decorate(THttpService.newDecorator()));
+ *
+ * }</pre>
+ *
+ * <p>Note that this class adds {@link HttpHeaders} only when the {@link HttpRequest} is decoded to
+ * an {@link RpcRequest} successfully in {@link THttpService}. If you want to add {@link HttpHeaders}
+ * regardless of the decoding succeeded or not, use your own decorator:
+ *
+ * <pre>{@code
+ * > public final class MyOwnInjector extends SimpleDecoratingService<HttpRequest, HttpResponse> {
+ * >     MyOwnInjector(Service<HttpRequest, HttpResponse> delegate) {
+ * >         super(delegate);
+ * >     }
+ * >
+ * >     @Override
+ * >     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+ * >         final HttpResponse originalRes = delegate().serve(ctx, req);
+ * >         final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+ * >         final HttpResponse newRes = HttpResponse.from(future);
+ * >
+ * >         originalRes.aggregate().whenComplete((res, cause) -> {
+ * >             if (cause != null) {
+ * >                 future.completeExceptionally(cause);
+ * >             } else {
+ * >                 final HttpHeaders headers =
+ * >                 res.headers().add(AsciiString.of("my-header"), "foo");
+ * >                 future.complete(HttpResponse.of(headers, res.content(), res.trailingHeaders()));
+ * >             }
+ * >         });
+ * >         return newRes;
+ * >     }
+ * > }
+ * }</pre>
+ *
+ * <p>Then, decorate it after {@link THttpService} is decorated:
+ *
+ * <pre>{@code
+ * final ServerBuilder sb = ...
+ * sb.service("/", ThriftCallService.of(thriftHandler)
+ *                                  .decorate(...)
+ *                                  .decorate(THttpService.newDecorator())
+ *                                  .decorate(MyOwnInjector::new));
+ * }</pre>
+ */
+public abstract class THttpHeaderInjectingService extends SimpleDecoratingService<RpcRequest, RpcResponse> {
```

Comment:
How about `HttpHeaderInjectingRpcService`? It should work for any `Service<RpcRequest, RpcResponse>`.